lib/Transforms/MemoryBanking.cpp

```diff
@@ -695,55 +757,208 @@ LogicalResult cleanUpOldMemRefs(DenseSet<Value> &oldMemRefVals,
+                                    originalDef->getAttr(bankingDimensionsStr));
+      } else if (isa<BlockArgument>(memref)) {
+        auto blockArg = cast<BlockArgument>(memref);
+        auto *parentOp = blockArg.getOwner()->getParentOp();
+        auto funcOp = dyn_cast<func::FuncOp>(parentOp);
+        assert(funcOp &&
+               "Expected the original memory to be a FuncOp block argument!");
+        unsigned argIndex = blockArg.getArgNumber();
+        SmallVector<unsigned, 4> specifiedOrDefaultDims =
+            getSpecifiedOrDefaultBankingDim(
+                bankingDimensions, memrefType.getRank(), memrefType.getShape());
+
+        if (!funcOp.getArgAttr(argIndex, bankingFactorsStr))
+          funcOp.setArgAttr(argIndex, bankingFactorsStr, defaultFactorsAttr);
+        if (!funcOp.getArgAttr(argIndex, bankingDimensionsStr))
+          funcOp.setArgAttr(argIndex, bankingDimensionsStr,
+                            getDimensionsAttr(specifiedOrDefaultDims));
+
+        verifyBankingAttributesSize(
+            funcOp.getArgAttr(argIndex, bankingFactorsStr),
+            funcOp.getArgAttr(argIndex, bankingDimensionsStr));
+      }
+      return WalkResult::advance();
+    });
+  });
+}
+
 void MemoryBankingPass::runOnOperation() {
-  if (getOperation().isExternal() || bankingFactor == 1)
+  this->bankingFactors = {bankingFactorsList.begin(), bankingFactorsList.end()};
+  this->bankingDimensions = {bankingDimensionsList.begin(),
+                             bankingDimensionsList.end()};
+
+  if (getOperation().isExternal() ||
+      (bankingFactors.empty() ||
+       std::all_of(bankingFactors.begin(), bankingFactors.end(),
+                   [](unsigned f) { return f == 1; })))
     return;
 
-  if (bankingFactor == 0) {
+  if (std::any_of(bankingFactors.begin(), bankingFactors.end(),
+                  [](int f) { return f == 0; })) {
     getOperation().emitError("banking factor must be greater than 1");
     signalPassFailure();
     return;
   }
 
-  getOperation().walk([&](mlir::affine::AffineParallelOp parOp) {
-    DenseSet<Value> memrefsInPar = collectMemRefs(parOp);
+  if (bankingDimensions.size() > bankingFactors.size()) {
+    getOperation().emitError(
+        "A banking dimension must be paired with a factor");
+    signalPassFailure();
+    return;
+  }
+  // `bankingFactors` is guaranteed to have elements and at least one of them is
+  // greater than 1 beyond this point.
+
+  setAllBankingAttributes(getOperation(), &getContext());
+
+  OpBuilder builder(getOperation());
+  // We run this pass until convergence, i.e., `applyMemoryBanking` has reached
+  // its fixed point, which means every memory read/write operation has been
+  // rewritten to be using the newly created banks, and that the old memory
+  // references are erased.
+  bool banksCreated = false;
+  do {
+    memoryToBanks.clear();
+    oldMemRefVals.clear();
+    opsToErase.clear();
+
+    banksCreated = false;
```

Comment:
Can't we delete this line?

Reply:
we should not right?
say a memory reference needs to be banked in multiple dimensions.
In the first pass to bank the 0'th dimension, `banksCreated` will be set to `true` after `createBanks`. Afterwards, `banksCreated` will remain `true`, so there's no way we can break out of the loop.

Reply:
Ah yes, you're definitely right.